python/lsst/ts/ledprojector/led_controller.py

```diff
@@ -0,0 +1,576 @@
+    log : `logging.Logger` or 'None', optional
+        Logger.
+    simulate : `bool`, optional
+        Run in simulation mode?
+    make_connect_time_out : `bool`, optional
+        Make the connect method timeout?
+        Only useful for unit tests.
+        Ignored if simulate false.
+    """
+
+    def __init__(
+        self,
+        config: types.SimpleNamespace,
+        log: logging.Logger | None = None,
+        simulate: bool = False,
+        make_connect_time_out: bool = False,
+    ) -> None:
+        super().__init__(
+            config=config, topics=config.topics, log=log, simulation_mode=simulate
+        )
+
+        self.config = config
+        self.log = (
+            log.getChild(type(self).__name__)
+            if log is not None
+            else logging.getLogger(type(self).__name__)
+        )
+        self.make_connect_time_out = make_connect_time_out
+
+        # We want a multiple key -> 1 value list for easy parsing
+        # We are creating a dictionary with 3 keys -> 1 value
+        # 1st key : Serial number of the LED
+        # 2nd key : 0-indexed list placement of the LED
+        # 3rd key : Labjack Channel Name of that LED
+        # First extract info from the config file,
+        # note that were going to use the first
+        # available 'ledControllerItem' in config,
+        # as there should only be one anyway
+        led_topic = next(
+            topic
+            for topic in self.config.topics
+            if topic["topic_name"] == "ledControllerItem"
+        )
+
+        # Extract the serial and lbj channel name lists
+        led_names = led_topic["led_names"]
+        channel_names = led_topic["channel_names"]
+
+        self.log.info(f"Opening led_controller with led names {led_names}")
+        self.log.info(f"Opening led_controller with channel names {channel_names}")
+
+        # combine the two lists into a list of tuples
+        partial_key_tuple = list(zip(led_names, channel_names))
+
+        # Now make a list of labjackchannel objs using the previous list's info
+        # Note that we use enumerate to get the numerical index
+        full_key_tuple = [
+            LabjackChannel(serial_number=data[0], channel=data[1], index=ind)
+            for ind, data in enumerate(partial_key_tuple, start=0)
+        ]
+        # Now create our multi-key dictionary
+        self.channels: Dict[Union[str, int], LabjackChannel] = {}
+        for lbc in full_key_tuple:
+            self.channels.update({lbc.serial: lbc})
+            self.channels.update({lbc.index: lbc})
+            self.channels.update({lbc.channel: lbc})
+
+        self.log.info(f"led_controller dictionary {self.channels}")
+
+        # Set if connected to the labjack and state data seen,
+        # cleared otherwise.
+        self.status_event = asyncio.Event()
+        self.status_task = utils.make_done_future()
+
+    @classmethod
+    def get_config_schema(cls) -> dict[str, Any]:
```

Comment:
It doesn't really matter to me, but I usually see this as it's own file rather than imbedded in the program. It might be easier to find if it is extracted into a different file.

Reply:
Usually we only split these for CSCs. In this particular case it is more common to do what Joseph did here.

Reply:
Ok, works for me.